### PR TITLE
CMake changes

### PR DESCRIPTION
**Pull request by snnn**
**Description**: 

1. Add support for vstest. 
2. Add support for vcpkg. To use it:
  ```bat
   vcpkg install zlib:x64-windows benchmark:x64-windows gtest:x64-windows protobuf:x64-windows pybind11:x64-windows re2:x64-windows
   mkdir build
   cmake ..\cmake -DCMAKE_BUILD_TYPE=Debug -A x64 -T host=x64 -DCMAKE_TOOLCHAIN_FILE=C:\vcpkg\scripts\buildsystems\vcpkg.cmake -DVCPKG_TARGET_TRIPLET=x64-windows -Donnxruntime_PREFER_SYSTEM_LIB=ON
  ```
3. New cmake option: onnxruntime_PREFER_SYSTEM_LIB, which allows user using the preinstall libs instead of the things in onnxruntime submodule.
4. New cmake option: onnxruntime_ENABLE_MEMLEAK_CHECKER, which allows user turn on/off the memory leak checker by @RyanUnderhill in Windows Debug Build. The checker doesn't work with vstest.
4. Fix the post merge pipeline(Mainly for test coverage report).
5. Ignore the compile warning from the Featurizer library code
6. Apply "/utf-8" VC compile flag to our code. Without this, you can't build onnxruntime on Chinese Windows.
7. Remove the SingleUnitTestProject cmake option because it's deprecated more than one year and nobody is using it.
8. Move opaque api tests to onnxruntime_test_all
9. Enable "/W4" on CUDA ep's C++ code(Not the *.cu files), and fix some warnings, add some extra checks.
10. Delete the onnxruntime::test::TestEnvironment class.
11. Add a DLLmain for onnxruntime.dll. 
12. Allow dynamic link to libprotobuf

**Motivation and Context**
- Why is this change required? What problem does it solve?
Related to #931

- If it fixes an open issue, please link to the issue here.

### CODE REVIEW

File: cmake/CMakeLists.txt [hunk 1]
```diff
@@ -135,11 +148,11 @@ if(onnxruntime_ENABLE_LTO)
       message(WARNING "IPO is not supported by this compiler")
       set(onnxruntime_ENABLE_LTO OFF)
     else()
+      set (CMAKE_INTERPROCEDURAL_OPTIMIZATION ON)
       if(UNIX)
         set (CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -flto")
         set (CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -flto")
         set (CMAKE_LDFLAGS "${CMAKE_CXX_FLAGS} -flto")
-        set (CMAKE_INTERPROCEDURAL_OPTIMIZATION ON)
```
Comment (tiagoshibata) [inline cmake/CMakeLists.txt]: Why do you need both `CMAKE_INTERPROCEDURAL_OPTIMIZATION` and setting `-flto` manually?

Comment (snnn) [reply cmake/CMakeLists.txt]: Good Question!

Comment (snnn) [reply cmake/CMakeLists.txt]: I'll remove it.

**Comment by pranavsharma** [comment]
Do we need to setup some pipeline or continuous test to ensure vcpkg support always works?

**Comment by pranavsharma** [comment]
Also, these are a lot of changes. Would've been nice to split them.

**Comment by snnn** [comment]
> Do we need to setup some pipeline or continuous test to ensure vcpkg support always works?

Good suggestion! I will do it.

**Pull request merged**